crucible/src/Lang/Crucible/CFG/SSAConversion.hs

```diff
@@ -377,46 +398,79 @@ inferBlockInfo blocks = seq input_map $ go bi0 blocks
+          case inferRegAssignment inputs of
+            Some ra -> do
+              let crepr = fmapFC typeOfValue ra
+              case untyped_id of
+                LabelID l -> do
                   let block_id = C.BlockID (nextIndex sz)
+                  let block_term = (blockTerm b) { pos_val = Right $ pos_val $ blockTerm b }
                   let binput = BInput { binputID = block_id
                                       , binputArgs    = ra
                                       , binputStmts   = blockStmts b
-                                      , binputTerm    = blockTerm b
+                                      , binputTerm    = block_term
                                       }
-                  let blocks' = extBlockInputAssignment $ biBlocks bi
-                  let jump_info'   = extJumpInfoMap $ biJumpInfo bi
-                  let switch_info' = extSwitchInfoMap $ biSwitchInfo bi
+                  let bi' = extBlockInfo bi binput
                   let ji = JumpInfo block_id crepr ra
-                  let bi' = BI { biBlocks = extend blocks' binput
-                               , biJumpInfo = insertJumpInfo l ji jump_info'
-                               , biSwitchInfo = switch_info'
-                               }
-                  go bi' rest
-            LambdaID l -> do
-              case inferRegAssignment inputs of
-                Some ra -> do
-                  let crepr = fmapFC typeOfValue ra
+                  let bi'' = bi' { biJumpInfo = insertJumpInfo l ji (biJumpInfo bi') }
+                  gogo bi'' rest
+                LambdaID l -> do
                   let block_id = C.BlockID (nextIndex sz)
                   let lastArg = AtomValue (lambdaAtom l)
+                  let block_term = (blockTerm b) { pos_val = Right $ pos_val $ blockTerm b }
                   let binput = BInput { binputID = block_id
                                       , binputArgs = ra :> lastArg
                                       , binputStmts = blockStmts b
-                                      , binputTerm = blockTerm b
+                                      , binputTerm = block_term
                                       }
-                  let blocks' = extBlockInputAssignment $ biBlocks bi
-                  let jump_info'   = extJumpInfoMap $ biJumpInfo bi
-                  let switch_info' = extSwitchInfoMap $ biSwitchInfo bi
+                  let bi' = extBlockInfo bi binput
                   let si = SwitchInfo block_id crepr ra
-                  let bi' = BI { biBlocks = extend blocks' binput
-                               , biJumpInfo = jump_info'
-                               , biSwitchInfo = insertSwitchInfo l si switch_info'
-                               }
-                  go bi' rest
+                  let bi'' = bi' { biSwitchInfo = insertSwitchInfo l si (biSwitchInfo bi') }
+                  gogo bi'' rest
+        gogo :: BlockInfo ext s ret blocks
```

Comment:
More than one `go` function probably means we need better names for both of them...